tutorials/securing-your-app-with-spring-security/04__form-less/content.adoc

```diff
@@ -0,0 +1,141 @@
+		if (requestURI != null && !requestURI.isEmpty() && !requestURI.contains(LoginView.ROUTE)) { // <2>
+			return requestURI.startsWith("/") ? requestURI.substring(1) : requestURI; // <3>
+		}
+	}
+
+	// if everything fails, redirect to the main view
+	return "";
+}
+----
+<1> Here we use our custom request cache to get the saved redirect URL.
+<2> As the request is only saved in exceptional cases (user is not logged in) by Spring Security, we make sure we do not run into a NPE here. Additionally, we prevent redirects to the login view.
+<3> Some mangling to satisfy `Ui.navigate()` that expects relative links without a leading slash.
+
+Finally, we will put everything together in the login view itself:
+
+.`*LoginView.java*`
+[source,java,linenums]
+----
+@Tag("sa-login-view")
+@Route(value = LoginView.ROUTE)
+@PageTitle("Login")
+public class LoginView extends VerticalLayout {
+    public static final String ROUTE = "login";
+
+    private LoginOverlay login = new LoginOverlay(); // <1>
+
+    @Autowired
+    public LoginView(AuthenticationManager authenticationManager, // <2>
+                     CustomRequestCache requestCache) {
+        // configures login dialog and adds it to the main view
+        login.setOpened(true);
+        login.setTitle("Spring Secured Vaadin");
+        login.setDescription("Login Overlay Example");
+
+        add(login);
+
+        login.addLoginListener(e -> { // <3>
+            try {
+                // try to authenticate with given credentials, should always return !null or throw an {@link AuthenticationException}
+                final Authentication authentication = authenticationManager
+                    .authenticate(new UsernamePasswordAuthenticationToken(e.getUsername(), e.getPassword())); // <4>
+
+                // if authentication was successful we will update the security context and redirect to the page requested first
+                SecurityContextHolder.getContext().setAuthentication(authentication); // <5>
+                login.close(); // <6>
+                UI.getCurrent().navigate(requestCache.resolveRedirectUrl()); // <7>
+
+            } catch (AuthenticationException ex) { // <8>
+		// show default error message
+		// Note: You should not expose any detailed information here like "username is known but password is wrong"
+		// as it weakens security.
+                login.setError(true);
+            }
+        });
+    }
+}
+----
+<1> Let's use the awesome login dialog component Vaadin provides.
+<2> Inject needed beans. The `authenticationManager` to kindly ask for login validation and the `requestCache` to get the redirect URL.
+<3> The component allows registering a login listener that gives access to the provided username and password.
+<4> Starts the authentication process by creating an authentication request object and let the manager do the rest. If successful we get a fully configured authentication object.
```

Comment:
ok problem with "let" here because it follows "and" but there are two verbs before it "starts" and "creating". So you need to decide, you have to have "let" as "lets" or "letting" depending on which verb (starts or creating) you want to compare with